ui/src/app/pages/analysis/notebook-redirect.tsx

```diff
@@ -0,0 +1,425 @@
+          if (creatingNewNotebook) {
+            return 'Creating the new notebook';
+          } else {
+            return 'Copying the notebook onto the server';
+          }
+        case 3:
+          return 'Redirecting to the notebook server';
+      }
+    };
+
+    // The last icon (Redirect) should be rotated 90 degrees
+    const rotateIcon = () => {
+      return icon === 'circle-arrow' ? 'rotate(90deg)' : undefined;
+    };
+
+    return <div style={isCurrent ? {...styles.progressCard, backgroundColor: '#F2FBE9'} :
+      styles.progressCard}>
+      {isCurrent ? <Spinner style={{width: '46px', height: '46px'}}
+                            data-test-id={'progress-card-spinner-' + index}/> :
+        <React.Fragment>
+          {icon === 'notebook' ? <NotebookIcon style={styles.progressIcon}/> :
+          <ClrIcon shape={icon} style={isComplete ?
+          {...styles.progressIcon, ...styles.progressIconDone,
+            transform: rotateIcon()} :
+            {...styles.progressIcon, transform: rotateIcon()}}/>}
+        </React.Fragment>}
+        <div style={styles.progressText}>
+          {renderText()}
+        </div>
+    </div>;
+  };
+
+interface State {
+  creatingNewNotebook: boolean;
+  notebookName: string;
+  fullNotebookName: string;
+  initialized: boolean;
+  leoUrl: string;
+  localizationError: boolean;
+  playgroundMode: boolean;
+  progress: Progress;
+  progressComplete: Map<Progress, boolean>;
+}
+
+interface Props {
+  workspace: WorkspaceData;
+  queryParams: any;
+  profileState: {profile: Profile, reload: Function, updateCache: Function};
+}
+
+export const NotebookRedirect = fp.flow(withUserProfile(), withCurrentWorkspace(),
+  withQueryParams())(class extends React.Component<Props, State> {
+
+    private timeoutReference: NodeJS.Timer;
+
+    constructor(props) {
+      super(props);
+      this.state = {
+        creatingNewNotebook: !!props.queryParams.creating,
+        fullNotebookName: undefined,
+        initialized: false,
+        leoUrl: undefined,
+        localizationError: false,
+        notebookName: undefined,
+        playgroundMode: props.queryParams.playgroundMode === true,
+        progress: Progress.Unknown,
+        progressComplete: new Map<Progress, boolean>(),
+      };
+    }
+
+    componentDidMount() {
+      this.setNotebookNames();
+      this.pollCluster(this.props.workspace.namespace);
+    }
+
+    componentWillUnmount() {
+      clearTimeout(this.timeoutReference);
+    }
+
+    isClusterInProgress(cluster: Cluster): boolean {
+      return cluster.status === ClusterStatus.Starting ||
+        cluster.status === ClusterStatus.Stopping ||
+        cluster.status === ClusterStatus.Stopped;
+    }
+
+    clearAndSetTimeout(timeoutCallback, timeoutMillis) {
+      clearTimeout(this.timeoutReference);
+      this.timeoutReference = setTimeout(() => timeoutCallback(), timeoutMillis);
+    }
+
+    async pollCluster(billingProjectId) {
+      const repoll = () => {
```

Comment:
I don't think we need to carry over this repolling pattern. As currently written, this is more blunt than the Angular implementation in terms of error handling. Since we're in an async function, we can rewrite this method in a more straight-line fashion. Here's what I would suggest:

```
while (true) {
  const {cluster} = await retry(() => clusterApi().listClusters(), 10000, 5);
  ...
  if (cluster.status is RUNNING) {
    break;
  }
  // Effectively sleep 15s
  await new Promise(resolve => setTimeout(resolve, 15000));
}

await retry(() => this.initializeCookies(), 5000, 3);
...


async retry<T>(f: () => Promise<T>, backoff: number, maxRetries: number): T {
  let retries = 0;
  while (true) {
    try {
      return await f();
    } catch (e) {
      retries++;
      if (isAbortError(e) || retries >= maxRetries) throw e;
    }
    await new Promise(resolve => setTimeout(resolve, backoff));
  }
}
```

Would also need to address timeout leakage for such an implementation. retry accepting the `AbortSignal` parameter would be pretty neat.

Reply:
This retry pattern exists in lots of places, so this would be a good candidate for a tech debt ticket.  I can remove it here.

Reply:
I believe I have address this now.